packages/codebytes/src/__tests__/codebyte-test.tsx

```diff
@@ -0,0 +1,197 @@
+  }
+  if (language) {
+    url.searchParams.set('lang', language);
+  }
+  url.searchParams.set('client-name', 'forum');
+  url.searchParams.set(
+    'page',
+    'https://discuss.codecademy.com/some-interesting/post'
+  );
+  if (mode) {
+    url.searchParams.set('mode', mode);
+  }
+  window.history.replaceState({}, '', url.toString());
+
+  return renderWrapper(rest);
+};
+
+describe('CodeBytes', () => {
+  const initialUrl = window.location.href;
+
+  afterEach(() => {
+    window.history.replaceState(null, '', initialUrl);
+    (trackClick as any).mockReset();
+    (trackUserImpression as any).mockReset();
+  });
+
+  it('has a language-specific "hello world" program defined for each language', () => {
+    validLanguages.forEach((language) => {
+      expect(helloWorld[language]).toBeDefined();
+    });
+  });
+
+  it('initializes with a language-specific "hello world" program when there is no language prop', () => {
+    const { view } = renderWrapper();
+    const selectedLanguage = view.getByRole('combobox') as Element;
+    userEvent.selectOptions(selectedLanguage, ['javascript']);
+    view.getByText(helloWorld.javascript);
+  });
+
+  it('initializes with a language-specific "hello world" program when there is a language prop but no text prop', () => {
+    const { view } = renderWrapper({ language: 'javascript' });
+    view.getByText(helloWorld.javascript);
+  });
+
+  it('initializes with deserialized text when there is a text prop but no language prop', () => {
+    const testString = 'yes hello';
+    const { view } = renderWrapper({ text: testString });
+    const selectedLanguage = view.getByRole('combobox') as Element;
+    userEvent.selectOptions(selectedLanguage, ['javascript']);
+    view.getByText(testString);
+  });
+
+  it('initializes with deserialized text when there is both a language and text prop', () => {
+    const testString = 'yes hello';
+    const { view } = renderWrapper({
+      text: testString,
+      language: 'javascript',
+    });
+    view.getByText(testString);
+  });
+
+  describe('Change Handlers', () => {
+    it('triggers onEdit on text edit', () => {
+      const onEdit = jest.fn();
+      const { view } = renderWrapper({
+        text: '',
+        language: 'javascript',
+        onEdit,
+      });
+
+      const editor = view.getByTestId(mockEditorTestId);
+      userEvent.type(editor, 'dog');
+
+      expect(onEdit).toHaveBeenCalledTimes(3);
+      expect(onEdit).toHaveBeenLastCalledWith('dog', 'javascript');
+    });
+
+    it('triggers onLanguageChange on language selection', () => {
+      const onLanguageChange = jest.fn();
+      const { view } = renderWrapper({
+        onLanguageChange,
+      });
+
+      const selectedLanguage = view.getByRole('combobox') as Element;
+      userEvent.selectOptions(selectedLanguage, ['javascript']);
+
+      expect(onLanguageChange).toHaveBeenCalledWith(
+        "console.log('Hello world!');",
+        'javascript'
+      );
+    });
+  });
```

Comment:
We still need a couple change handler functions, because these differ in the forums version and the regular codebytes. Here we can pass custom functions into the changeHandlers for forums as needed.